rust/src/lib.rs

```diff
@@ -1351,51 +2116,60 @@ pub extern "C" fn zcashlc_shield_funds(
 ) -> i64 {
     let res = catch_panic(|| {
         let network = parse_network(network_id)?;
-        let db_data = wallet_db(db_data, db_data_len, network)?;
+        let db_data = unsafe { wallet_db(db_data, db_data_len, network)? };
         let mut update_ops = (&db_data)
             .get_update_ops()
             .map_err(|e| format_err!("Could not obtain a writable database connection: {}", e))?;
 
-        let account = if account >= 0 {
-            account as u32
+        let usk = unsafe { decode_usk(usk_ptr, usk_len) }?;
+
+        let memo_bytes = if memo.is_null() {
+            MemoBytes::empty()
         } else {
-            return Err(format_err!("account argument must be positive"));
+            MemoBytes::from_bytes(unsafe { slice::from_raw_parts(memo, 512) })
+                .map_err(|e| format_err!("Invalid MemoBytes {}", e))?
         };
-        let tsk_wif = unsafe { CStr::from_ptr(tsk) }.to_str()?;
-        let extsk = unsafe { CStr::from_ptr(extsk) }.to_str()?;
-        let memo = unsafe { CStr::from_ptr(memo) }.to_str()?;
+
         let spend_params = Path::new(OsStr::from_bytes(unsafe {
             slice::from_raw_parts(spend_params, spend_params_len)
         }));
         let output_params = Path::new(OsStr::from_bytes(unsafe {
             slice::from_raw_parts(output_params, output_params_len)
         }));
 
-        //grab secret private key for t-funds
-        let sk: SecretKey = (&Wif(tsk_wif.to_string())).try_into()?;
+        let account = db_data
+            .get_account_for_ufvk(&usk.to_unified_full_viewing_key())?
+            .ok_or_else(|| format_err!("Spending key not recognized."))?;
 
-        let extsk =
-            match decode_extended_spending_key(network.hrp_sapling_extended_spending_key(), &extsk)
-            {
-                Ok(Some(extsk)) => extsk,
-                Ok(None) => {
-                    return Err(format_err!("ExtendedSpendingKey is for the wrong network"));
-                }
-                Err(e) => {
-                    return Err(format_err!("Invalid ExtendedSpendingKey: {}", e));
-                }
-            };
+        let taddrs: Vec<TransparentAddress> = db_data
+            .get_target_and_anchor_heights(0u32)
+            .map_err(|e| format_err!("Error while fetching anchor height: {}", e))
+            .and_then(|opt_anchor| {
+                opt_anchor
+                    .map(|(h, _)| h)
```

Comment:
Given that this is named `opt_anchor`, this extracts the wrong value; the tuple is `(target, anchor)`. Then again, with a confirmations of `0`, the two should be identical.

Reply:
There are two other occurrences of `opt_anchor` that then extract the target height instead of the anchor height. Are those also incorrect?

Reply:
Probably yes. Can you give links?

Reply:
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1253-L1255
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1307-L1309
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1376-L1378
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1428-L1430